### PR TITLE
refactor: generalize msg access control

### PR DESCRIPTION
**Pull request by cgorenflo**

### CODE REVIEW

**Review by haiyizxx — APPROVED** [review]

File: contracts/router/src/contract.rs [hunk 1]
```diff
@@ -67,59 +71,49 @@ pub fn execute(
...
+                RouterStore::new(deps.storage),
+                info.sender,
+                msgs,
+            )?)
+        }
     }
     .map_err(axelar_wasm_std::ContractError::from)
 }
 
-struct Contract<S>
```
Comment (haiyizxx) [inline contracts/router/src/contract.rs:103]: why remove the struct?

File: packages/axelar-wasm-std/src/permission_control.rs [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+use cosmwasm_std::{Addr, StdResult};
+use cw_storage_plus::Item;
+use flagset::flags;
+use serde::{Deserialize, Serialize};
+
+flags! {
+    #[repr(u8)]
+    #[derive(Serialize, Deserialize)]
+    pub enum Permission: u8 {
+        NoPrivilege = 1, // this specifies that the user MUST NOT have an elevated role
```
Comment (cjcobb23) [inline packages/axelar-wasm-std/src/permission_control.rs:11]: What's the usecase for this?

File: packages/axelar-wasm-std/src/permission_control.rs [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+        Governance = 4,
+        Elevated = (Permission::Admin | Permission::Governance).bits(),
+        Any = (Permission::NoPrivilege | Permission::Elevated).bits(),
+    }
+}
+
+#[derive(thiserror::Error, Debug, PartialEq)]
+pub enum Error {
+    #[error("sender is not allowed to perform this action")]
+    PermissionDenied,
```
Comment (cjcobb23) [inline packages/axelar-wasm-std/src/permission_control.rs:22]: Maybe add the privilege level required to the error message?

File: packages/axelar-wasm-std/src/permission_control.rs [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+
+const GOVERNANCE: Item<Addr> = Item::new("permission_control_governance_addr");
+
+pub fn set_admin(storage: &mut dyn cosmwasm_std::Storage, addr: &Addr) -> StdResult<()> {
+    ADMIN.save(storage, addr)
+}
+
+pub fn set_governance(storage: &mut dyn cosmwasm_std::Storage, addr: &Addr) -> StdResult<()> {
+    GOVERNANCE.save(storage, addr)
+}
```
Comment (cjcobb23) [inline packages/axelar-wasm-std/src/permission_control.rs:98]: How does this work with multiple contracts? Do they each have their own storage, and the db somehow prefixes by contract address or some such?

File: packages/axelar-wasm-std/src/permission_control.rs (hunk 2 above)
Comment (cgorenflo) [reply packages/axelar-wasm-std/src/permission_control.rs]: no use case on it's own, I need it to define the `Any` case

File: packages/axelar-wasm-std/src/permission_control.rs (hunk 4 above)
Comment (cgorenflo) [reply packages/axelar-wasm-std/src/permission_control.rs:131]: yes, the storage is namespaced to the contract

File: contracts/router/src/contract.rs [hunk 5]
```diff
@@ -20,10 +23,7 @@ pub fn migrate(
     _env: Env,
     _msg: Empty,
 ) -> Result<Response, axelar_wasm_std::ContractError> {
-    // any version checks should be done before here
-
-    cw2::set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
-
```
Comment (cjcobb23) [inline contracts/router/src/contract.rs:26]: Are we setting the version somewhere else now?

Comment (cgorenflo) [reply contracts/router/src/contract.rs:26]: hmm, that line got lost. Added it back in

**Pull request merged**